### PR TITLE
Create StateFactory

### PR DESCRIPTION
**Pull request by danieljanes**

### CODE REVIEW

File: src/py/flwr/server/state/state_factory.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+    def __init__(self, database: str) -> None:
+        self.database = database
+        self.state_instance: Optional[State] = None
+
+    def state(self) -> State:
+        """Return a State instance and create it, if necessary."""
+
+        # InMemoryState
+        if self.database == ":flwr-in-memory-state:":
```
Comment (tanertopal) [inline src/py/flwr/server/state/state_factory.py:39]:
```suggestion
        if self.database == ":flwr-in-memory-state:":
```
This format is valid for SqliteState, too. Is that intentional?

File: src/py/flwr/server/state/state_factory.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            if self.state_instance is None:
+                self.state_instance = InMemoryState()
+            log(DEBUG, "Using InMemoryState")
+            return self.state_instance
+
+        # SqliteState
+        state = SqliteState(self.database)
+        state.initialize()
```
Comment (tanertopal) [inline src/py/flwr/server/state/state_factory.py:47]:
```suggestion
        state = SqliteState(self.database)
        state.initialize()
```
Just a remark: I already had a draft of a threading-compatible version of the SqliteState. Let's not build too many workarounds around it. I assume that's the intention here, as it seems easily adjustable with the `self.state_instance` property.

File: src/py/flwr/server/state/state_factory.py (hunk 1 above)
Comment (danieljanes) [reply src/py/flwr/server/state/state_factory.py:39]: It's inspired by it, but we could use any magic string here.

File: src/py/flwr/server/state/state_factory.py (hunk 2 above)
Comment (danieljanes) [reply src/py/flwr/server/state/state_factory.py:47]: Exactly, a different kind of `State` implementation would only require changes in the `StateFactory.state` method.

File: src/py/flwr/server/state/state_factory.py (hunk 1 above)
Comment (danieljanes) [reply src/py/flwr/server/state/state_factory.py:39]: (btw, easy to change it later)

**Review by tanertopal — APPROVED** [review]

**Pull request merged**